### ISSUE TITLE
batch_normalization 'high_mem' vs 'low_mem'

### ISSUE THREAD

**Issue by lucasb-eyer** [Theano/Theano#4590]
So, I wanted to save some GPU memory and try out Theano's 'low_mem' batchnorm implementation, but it doesn't seem to use any less memory than the corresponding plain theano expression (i.e. 'high_mem') by looking at `nvidia-smi`.

I then looked at the Theano profiler, it even seems that 'high_mem' uses less memory? What gives? I'm slightly confused and this is the first time I use Theano's profiler, so please excuse any noob mistakes.

Test code:

``` python
import os
os.environ['CUDA_LAUNCH_BLOCKING'] = '1'
os.environ['THEANO_FLAGS'] = 'profile=True,profile_memory=True'

import sys

import numpy as np

import theano as th
import theano.tensor as T

x = T.matrix("x")
g, b = th.shared(np.float32(1)), th.shared(np.float32(0))
m = T.mean(x, axis=1, keepdims=True)
s = T.std(x, axis=1, keepdims=True)

X = np.random.randn(8192, 8192).astype(np.float32)

bn = T.nnet.batch_normalization(x, g, b, m, s, mode='high_mem' if sys.argv[1] == 'hi' else 'low_mem')
fn = th.function([x], bn)

for i in range(10):
    out = np.array(fn(X))

print(np.mean(out))
```

Output of `python memtest.py hi`, only the memory-relevant part:

```
Memory Profile
(Sparse variables are ignored)
(For values in brackets, it's for linker = c|py

---
    Max peak memory with current setting
        CPU: 262144KB (262144KB)
        GPU: 262176KB (262176KB)
        CPU + GPU: 524320KB (524320KB)
    Max peak memory with current setting and Theano flag optimizer_excluding=inplace
        CPU: 262144KB (262144KB)
        GPU: 524320KB (524320KB)
        CPU + GPU: 786464KB (786464KB)
    Max peak memory if allow_gc=False (linker don't make a difference)
        CPU: 262144KB
        GPU: 262208KB
        CPU + GPU: 524352KB

---

    <Sum apply outputs (bytes)> <Apply outputs shape> <created/inplace/view> <Apply node>

     268435456B  [(8192, 8192)] c HostFromGpu(GpuElemwise{Composite{((i0 * i1) + i2)}}[(0, 0)].0)
     268435456B  [(8192, 8192)] i GpuElemwise{Sub}[(0, 0)](GpuFromHost.0, GpuElemwise{TrueDiv}[(0, 0)].0)
     268435456B  [(8192, 8192)] c GpuFromHost(x)
     268435456B  [(8192, 8192)] i GpuElemwise{Composite{((i0 * i1) + i2)}}[(0, 0)](GpuElemwise{Sub}[(0, 0)].0, GpuElemwise{Composite{(i0 / sqrt((i1 / i2)))}}[(0, 1)].0, GpuDimShuffle{x,x}.0)
         32768B  [(8192,)] c GpuCAReduce{pre=sqr,red=add}{0,1}(GpuElemwise{Sub}[(0, 0)].0)
         32768B  [(8192,)] c GpuCAReduce{add}{0,1}(GpuFromHost.0)
         32768B  [(8192, 1)] i GpuElemwise{Composite{(i0 / sqrt((i1 / i2)))}}[(0, 1)](GpuDimShuffle{x,x}.0, GpuDimShuffle{0,x}.0, GpuFromHost.0)
         32768B  [(8192, 1)] v GpuDimShuffle{0,x}(GpuCAReduce{add}{0,1}.0)
         32768B  [(8192, 1)] v GpuDimShuffle{0,x}(GpuCAReduce{pre=sqr,red=add}{0,1}.0)
         32768B  [(8192, 1)] i GpuElemwise{TrueDiv}[(0, 0)](GpuDimShuffle{0,x}.0, GpuFromHost.0)
   ... (remaining 8 Apply account for   40B/1073938472B ((0.00%)) of the Apply with dense outputs sizes)

    <created/inplace/view> is taken from the Op's declaration.
    Apply nodes marked 'inplace' or 'view' may actually allocate memory, this is not reported here. If you use DebugMode, warnings will be emitted in those cases.
```

Output of `python memtest.py lo`, only the memory-relevant part:

```
Memory Profile
(Sparse variables are ignored)
(For values in brackets, it's for linker = c|py

---
    Max peak memory with current setting
        CPU: 262144KB (262144KB)
        GPU: 524352KB (524352KB)
        CPU + GPU: 786496KB (786496KB)
    Max peak memory with current setting and Theano flag optimizer_excluding=inplace
        CPU: 262144KB (262144KB)
        GPU: 524352KB (524352KB)
        CPU + GPU: 786496KB (786496KB)
    Max peak memory if allow_gc=False (linker don't make a difference)
        CPU: 262144KB
        GPU: 524352KB
        CPU + GPU: 786496KB

---

    <Sum apply outputs (bytes)> <Apply outputs shape> <created/inplace/view> <Apply node>

     268435456B  [(8192, 8192)] c HostFromGpu(GpuElemwise{Composite{((((i0 - i1) / i2) * i3) + i4)}}[(0, 0)].0)
     268435456B  [(8192, 8192)] c GpuFromHost(x)
     268435456B  [(8192, 8192)] i GpuElemwise{Composite{((((i0 - i1) / i2) * i3) + i4)}}[(0, 0)](GpuFromHost.0, GpuElemwise{TrueDiv}[(0, 0)].0, GpuElemwise{Composite{sqrt((i0 / i1))}}[(0, 0)].0, GpuDimShuffle{x,x}.0, GpuDimShuffle{x,x}.0)
     268435456B  [(8192, 8192)] c GpuElemwise{sub,no_inplace}(GpuFromHost.0, GpuElemwise{TrueDiv}[(0, 0)].0)
         32768B  [(8192,)] c GpuCAReduce{add}{0,1}(GpuFromHost.0)
         32768B  [(8192, 1)] v GpuDimShuffle{0,x}(GpuCAReduce{pre=sqr,red=add}{0,1}.0)
         32768B  [(8192,)] c GpuCAReduce{pre=sqr,red=add}{0,1}(GpuElemwise{sub,no_inplace}.0)
         32768B  [(8192, 1)] i GpuElemwise{TrueDiv}[(0, 0)](GpuDimShuffle{0,x}.0, GpuFromHost.0)
         32768B  [(8192, 1)] v GpuDimShuffle{0,x}(GpuCAReduce{add}{0,1}.0)
         32768B  [(8192, 1)] i GpuElemwise{Composite{sqrt((i0 / i1))}}[(0, 0)](GpuDimShuffle{0,x}.0, GpuFromHost.0)
   ... (remaining 8 Apply account for   40B/1073938472B ((0.00%)) of the Apply with dense outputs sizes)
```

**Comment by nouiz** [comment]
This is interresting. Can you share the code that reproduce this? Whih
version of Theano do you use? There was a fix in the dev version since the
last release in the peak memory computation.

**Comment by lucasb-eyer** [comment]
I'm slightly confused as the code is right in my post, maybe you missed it?

I'm on master from late March, I'll update and run again this evening but none of the commits affecting bn.py that have been made since March look like they'd affect performance.

**Comment by nouiz** [comment]
I would recommand to use cudnn batch norm. https://github.com/Theano/Theano/pull/5190 make a CPU inferface too, so closing.
